microcloud/service/microcloud.go

```diff
@@ -106,36 +100,34 @@ func (s CloudService) Join(ctx context.Context, joinConfig JoinConfig) error {
 	return s.client.JoinCluster(s.name, util.CanonicalNetworkAddress(s.address, s.port), joinConfig.Token, 5*time.Minute)
 }
 
-// RequestJoin notifies the peers that that should begin the join operation.
-func (s CloudService) RequestJoin(ctx context.Context, secrets map[string]string, joinConfig map[string]types.ServicesPut) chan joinResponse {
-	joinedChan := make(chan joinResponse, len(joinConfig))
-	for peer, cfg := range joinConfig {
-		go func(peer string, cfg types.ServicesPut) {
-			if secrets[peer] == "" {
-				joinedChan <- joinResponse{Name: peer, Error: fmt.Errorf("No auth secret found for peer")}
-				return
-			}
-
-			c, err := s.client.RemoteClient(util.CanonicalNetworkAddress(cfg.Address, CloudPort))
-			if err != nil {
-				joinedChan <- joinResponse{Name: peer, Error: err}
-				return
-			}
+// RequestJoin sends the signal to initiate a join to the remote system.
+// This can become blocking if there is a dqlite error, so instead return a channel immediately.
+func (s CloudService) RequestJoin(ctx context.Context, secret string, name string, joinConfig types.ServicesPut) chan error {
+	joinFunc := func() error {
+		c, err := s.client.RemoteClient(util.CanonicalNetworkAddress(joinConfig.Address, CloudPort))
+		if err != nil {
+			return err
+		}
 
-			c.Client.Client.Transport = &http.Transport{
-				TLSClientConfig:   &tls.Config{InsecureSkipVerify: true},
-				DisableKeepAlives: true,
-				Proxy: func(r *http.Request) (*url.URL, error) {
-					r.Header.Set("X-MicroCloud-Auth", secrets[peer])
+		c.Client.Client.Transport = &http.Transport{
+			TLSClientConfig:   &tls.Config{InsecureSkipVerify: true},
+			DisableKeepAlives: true,
+			Proxy: func(r *http.Request) (*url.URL, error) {
+				r.Header.Set("X-MicroCloud-Auth", secret)
 
-					return shared.ProxyFromEnvironment(r)
-				},
-			}
+				return shared.ProxyFromEnvironment(r)
+			},
+		}
 
-			joinedChan <- joinResponse{Name: peer, Error: client.JoinServices(ctx, c, cfg)}
-		}(peer, cfg)
+		return client.JoinServices(ctx, c, joinConfig)
 	}
 
+	joinedChan := make(chan error)
+
+	go func() {
+		joinedChan <- joinFunc()
```

Comment:
Can we accept a context rather than starting a go routine which could block forever? That way the caller can specify timeout and this function would return if context was canceled. Usually this sort of uncancellable go routine is a code smell to me.

Reply:
I had tried this initially with a context timeout, but it didn't work for some reason. If dqlite locked up on cluster join, the whole request would still block.

Looking at it again, it seems it's because I'm using `ConnectLXDUnix` rather than `ConnectLXDUnixWithContext` and `operation.Wait` instead of `operation.WaitContext` so the context wasn't actually propagating to the underlying request. 

That makes this a lot simpler after all, thanks!

Reply:
Excellent! You're welcome 😀